couchbase/io/streams.hxx

```diff
@@ -72,71 +76,83 @@ class stream_impl
 class plain_stream_impl : public stream_impl
 {
   private:
-    asio::ip::tcp::socket stream_;
+    std::shared_ptr<asio::ip::tcp::socket> stream_;
 
   public:
     explicit plain_stream_impl(asio::io_context& ctx)
       : stream_impl(ctx, false)
-      , stream_(strand_)
-    {
-    }
-
-    [[nodiscard]] bool is_open() const override
+      , stream_(std::make_unique<asio::ip::tcp::socket>(strand_))
     {
-        return stream_.is_open();
     }
 
     [[nodiscard]] asio::ip::tcp::endpoint local_endpoint() const override
     {
+        if (!open_ || !stream_) {
+            return {};
+        }
         std::error_code ec;
-        auto res = stream_.local_endpoint(ec);
+        auto res = stream_->local_endpoint(ec);
         if (ec) {
             return {};
         }
         return res;
     }
 
-    void close() override
+    void close(std::function<void(std::error_code)>&& handler) override
     {
-        stream_.close();
+        if (!open_ || !stream_) {
```

Comment:
I think this use of open_ and stream_ is not thread safe since it's not occurring in the strand.

Reply:
open_ is atomic_bool, so it should be safe

Reply:
If that's the case, then stream_ needs to be std::atomic<std::shared_ptr> in that case, and it's important that both open_ and stream_ are always checked together, since they are independently atomic.

Reply:
I removed this check, as `stream_` cannot be null and close could be called safely several times.